lib/core/utils/process-message.js

```diff
@@ -0,0 +1,85 @@
+/* global helpers */
+const dataRegex = /\$\{\s?data\s?\}/g;
+
+/**
+ * Replace a placeholder with the value of a data string or object
+ * @param {String} str
+ * @param {String|Object} data
+ */
+function substitute(str, data) {
+	// replace all instances of ${ data } with the value of the string
+	if (typeof data === 'string') {
+		return str.replace(dataRegex, data);
+	}
+
+	// replace all instances of ${ data[prop] } with the value of the property
+	for (let prop in data) {
+		if (data.hasOwnProperty(prop)) {
+			const regex = new RegExp('\\${\\s?data\\.' + prop + '\\s?}', 'g');
+			str = str.replace(regex, data[prop]);
+		}
+	}
+
+	return str;
+}
+
+/**
+ * Process a metadata message.
+ * @param {String|Object} message
+ * @param {Object} data
+ * @return {String}
+ */
+axe.utils.processMessage = function processMessage(message, data) {
```

Comment:
Is it worth pulling in one of the millions of maintained/battle-tested templating libraries for this rather than building our own?

Reply:
This was already discussed and approved in the proposal this is based on.